Mathlib/Data/Nat/BinaryRec.lean

```diff
@@ -0,0 +1,133 @@
+/-
+Copyright (c) 2022 Praneeth Kolichala. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Praneeth Kolichala, Yuyang Zhao
+-/
+
+/-!
+# Binary recursion on `Nat`
+
+This file defines binary recursion on `Nat`.
+
+## Main results
+* `Nat.binaryRec`: A recursion principle for `bit` representations of natural numbers.
+* `Nat.binaryRec'`: The same as `binaryRec`, but the induction step can assume that if `n=0`,
+  the bit being appended is `true`.
+* `Nat.binaryRecFromOne`: The same as `binaryRec`, but special casing both 0 and 1 as base cases.
+-/
+
+universe u
+
+namespace Nat
+
+/-- `bit b` appends the digit `b` to the little end of the binary representation of
+  its natural number input. -/
+def bit (b : Bool) (n : Nat) : Nat :=
+  cond b (n <<< 1 + 1) (n <<< 1)
```

Comment:
Presumably `| 1` is also faster than `+ 1`, if you're really micro-optimizing?

Reply:
This doesn't look like a good change to me. We have already settled on using `2 * n` and `2 * n + 1` in place of the old `bit0, bit1`, and these play better with ring theory tactics like `ring` and `linarith`. If you are doing this to improve the performance of the function, use a `@[csimp]` lemma instead.